fem/libceed/ceed.cpp

```diff
@@ -253,19 +348,68 @@ static void InitCeedTensorBasisAndRestriction(const FiniteElementSpace &fes,
                              tp_el_dof.GetData(), restr);
 }
 
+static void InitCeedTensorBasisAndRestriction(const FiniteElementSpace &fes,
+                                              const IntegrationRule &ir,
+                                              Ceed ceed, CeedBasis *basis,
+                                              CeedElemRestriction *restr)
+{
+   InitCeedTensorBasis(fes, ir, ceed, basis);
+   InitCeedTensorRestriction(fes, ir, ceed, restr);
+}
+
 void InitCeedBasisAndRestriction(const FiniteElementSpace &fes,
                                  const IntegrationRule &irm,
                                  Ceed ceed, CeedBasis *basis,
                                  CeedElemRestriction *restr)
 {
-   if (UsesTensorBasis(fes))
+   // Check for FES -> basis, restriction in hash tables
+   const Mesh *mesh = fes.GetMesh();
+   const FiniteElement *fe = fes.GetFE(0);
+   const int P = fe->GetDof();
+   const int Q = irm.GetNPoints();
+   const int nelem = mesh->GetNE();
+   std::string basis_key = std::to_string(reinterpret_cast<intptr_t>(&fes)) +
+                           std::to_string(reinterpret_cast<intptr_t>(&irm)) +
+                           std::to_string(P) + std::to_string(Q);
+   auto basis_itr = internal::ceed_basis_table.find(basis_key);
+   std::string restr_key = std::to_string(reinterpret_cast<intptr_t>(&fes)) +
+                           std::to_string(P) + std::to_string(nelem);
+   auto restr_itr = internal::ceed_restr_table.find(restr_key);
```

Comment:
I think it would be cleaner to use a tuple rather than doing the hash function manually each time you want to create a key, that's bug prone:
```c++
// For CeedBasis
using BasisKey = std::tuple<FiniteElementSpace*, IntegrationRule*, int, int>;
struct BasisHash
{
   std::size_t operator()(const BasisKey& k) const
   {
      return ...; // use std::get<n>(k) for computing the hash
   }
};
using BasisMap = std::unordered_map<const BasisKey, CeedBasis, BasisHash>;
auto basis_key = std::make_tuple(&fes,&irm,P,Q);

// For CeedRestriction
using RestrKey = std::tuple<FiniteElementSpace*, int, int>;
struct RestrHash
{
   std::size_t operator()(const RestrKey& k) const
   {
      return ...; // use std::get<n>(k) for computing the hash
   }
};
using RestrMap = std::unordered_map<const RestrKey, CeedRestriction, RestrHash>;
auto restr_key = std::make_tuple(&fes,P,nelem);
```

Reply:
I thought about that and its what I had at first. I just wasn't sure that we wanted to also build out the hashing and comparison operators.

This will, amusingly enough, result in me duplicating the code already included libCEED for hashing keys tuples.

Reply:
You don't need to build the comparison, only the hash doesn't have a default implementation for `tuple`.